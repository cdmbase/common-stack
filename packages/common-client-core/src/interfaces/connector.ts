// tslint:disable:max-line-length

import { LocalStateFragmentMatcher } from 'apollo-client';
import { IdGetter } from 'apollo-cache-inmemory';
import { ReducersMapObject } from 'redux';
import { ErrorLink } from 'apollo-link-error';
export interface IClientStateConfig {
    resolvers?: any; // don't need `Resolvers` type as it may conflict with the usage
    defaults?: any;
    typeDefs?: string | string[];
    fragmentMatcher?: LocalStateFragmentMatcher;
}


/**
 * ModuleShape have optional configuration to be implemented by all the feature modules
 * in the application.
 */
export interface IModuleShape {
    /**
     * @param link all `apollo-link` that need to be composed.
     * @inheritdoc https://github.com/apollographql/apollo-link
     */
    readonly link?: any;
    /**
     * @param errorLink compose all errorLink
     * @inheritdoc https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-error
     */
    readonly errorLink?: ErrorLink | ErrorLink[];
    readonly createFetch?: any;
    /**
     * @param connectionParam method that called when a client connects to the socket.
     * @inheritdoc https://github.com/apollographql/subscriptions-transport-ws#constructorurl-options-connectioncallback
     */
    readonly connectionParam?: any;
    /**
     * @param epic all `epics` that need to be composed.
     * @inheritdoc https://redux-observable.js.org/docs/basics/Epics.html
     */
    readonly epic?: any;
    /**
     * @deprecated not sure we using
     */
    readonly reduxContext?: any;
    /**
     * @param reducer Redux reducers list
     */
    readonly reducer?: ReducersMapObject | ReducersMapObject[];
    /**
     * @param clientStateParams Client side configuration of `apollo-client`
     * @inheritdoc https://github.com/apollographql/apollo-client/tree/2d65da133c156f6d808e64aee7e7fd5f7cc71d7f
     */
    readonly clientStateParams?: IClientStateConfig | IClientStateConfig[];
    /**
     * @param dataIdFromObject A function that returns an object identifier given an particular result
     * object.
     * @inheritdoc https://github.com/apollographql/apollo-client/tree/ed66999bac40226abfeada8d6c83b454636bb4b0/packages/apollo-cache-inmemory#configuration
     */
    readonly dataIdFromObject?: { [key: string]: IdGetter } |  { [key: string]: IdGetter }[];
    /**
     * @param createContainerFunc Synchronous Container Modules of inversify.
     * @inheritdoc https://github.com/inversify/InversifyJS/blob/master/wiki/container_modules.md
     */
    readonly createContainerFunc?: Function | Function[];
    readonly sidebarSegments?: any;
    readonly routerFactory?: any;
    /**
     * @param route Route list
     */
    readonly route?: any;
    /**
     * @param routeConfig Custom route data to create React Routers
     */
    readonly routeConfig?: any;
    /**
     * @param menuConfig Menu data to create side menu
     */
    readonly menuConfig?: any;
    /**
     * @param navItem Top left navigation links
     */
    readonly navItem?: any;
    /**
     * @param navItemRightTop right navigation links
     */
    readonly navItemRight?: any;
    /**
     * @param rootComponentFactory Root component factory list
     */
    readonly rootComponentFactory?: any;
    /**
     * @param dataRootComponent Data root React elements list (data root elements wraps data fetching react subtree root)
     */
    readonly dataRootComponent?: any;
    readonly createFetchOptions?: any;
    /**
     * @param stylesInsert URL list to 3rd party css scripts
     */
    readonly stylesInsert?: any;
    /**
     * @param scriptsInsert URL list to 3rd party js scripts
     */
    readonly scriptsInsert?: any;
    /**
     * @deprecated
     */
    readonly catalogInfo?: any;
    /**
     * @deprecated
     */
    readonly languagesFuncs?: any;
    /**
     * @param leftMainPanelItems Components that will be placed left panel.
     */
    readonly leftMainPanelItems?: any;
    /**
     * @param middleMainPanelItems Components that will be placed middle center panel
     */
    readonly middleMainPanelItems?: any;
    /**
     * @param middleMainPanelItemsProps Props of the component that will be placed middle center panel
     */
    readonly middleMainPanelItemsProps?: any;
    /**
     * @param leftFooterItems Components that will be placed left footer panel
     */
    readonly leftFooterItems?: any;
    /**
     * @param rightFooterItems Components that will be placed right footer panel
     */
    readonly rightFooterItems?: any;
    /**
     * @param middleLowerPanelItems Components that will be placed below middle panel
     */
    readonly middleLowerPanelItems?: any;

}

export interface IFeature extends IModuleShape {
    // Public variables
    readonly data: any[];

    /**
     * @returns Redux-Observable Epics
     */
    readonly epics;

    readonly getRouter;
    /**
     * @returns client-side React route components list
     */
    readonly getRoutes;
    readonly getConfiguredRoutes;

    readonly getMenus;
    readonly getConfiguredMenus;
    /**
     * @returns client-side top left navbar link component list
     */
    readonly navItems;
    /**
     * @returns client-side top right navbar link component list
     */
    readonly navItemsRight;
    /**
     * @returns Redux reducers
     */
    readonly reducers;

    readonly connectionParams;

    /**
     * @returns URL list to 3rd party css styles
     */
    readonly stylesInserts;
    /**
     * @returns URL list to 3rd party js styles
     */
    readonly scriptsInserts;

    readonly leftMainPanel;
    readonly middleMainPanel;
    readonly leftFooter;
    readonly rightFooter;
    readonly middleLowerPanel;
    readonly dataIdFromObject;

    /**
     * @param args Options to pass to each Container Module
     * @returns Created container and binds all the container modules
     */
    createContainers(args);
    /**
     * @param args Provide resolverContext which can be passed to all resolver functions.
     * @returns IClientStateConfig
     */
    getStateParams(args?: {resolverContex?: any }): IClientStateConfig;
    /**
     * @arguments root React tree root component
     * @arguments req Http Request
     * @returns React tree root component wrapped up by root components exposed by this module
     */
    getWrappedRoot(root: any, req?: any): any;
    /**
     * @param root React tree data root component (first React root components which is used for fetching data)
     *
     * @returns React tree data root component wrapped up by data root components exposed by this module
     */
    getDataRoot(root: any): any;
    registerLanguages(monaco): void;
}
