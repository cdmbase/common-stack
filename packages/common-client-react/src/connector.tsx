import * as React from 'react';

import { AbstractFeature, IFeature } from '@common-stack/client-core';


export class Feature extends AbstractFeature implements IFeature {
    get routes() {
        return this.route.map((component, idx) => React.cloneElement(component, { key: idx + this.route.length }));
    }

    get navItems() {
        return this.navItem.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemRight.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    public getWrappedRoot(root, req) {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

}
