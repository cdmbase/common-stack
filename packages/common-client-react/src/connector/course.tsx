import * as React from 'react';
import { Link } from 'react-router-dom';

interface MyProps {
  history?: any;
  children?: any;
  key?: any;
};

const Course: React.StatelessComponent<MyProps> = (props: MyProps) => {
  const from = "localhost:3000/"; 
  return (
  <div key={props.history.id}>
    <Link to={`${props.history.number}`} {...from}>{`Page: ${
        props.history.number
      }`}</Link>
  </div>
)};

export default Course;