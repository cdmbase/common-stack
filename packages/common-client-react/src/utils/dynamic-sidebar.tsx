import * as React from 'react';
import { Query, withApollo } from 'react-apollo';
import gql from 'graphql-tag';
import Course from '../connector/course';

export function renderQuery(loading, error, data) {
  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;
  return data.allPageHistories.map(allHistory => (
    <Course key={allHistory.id} history={allHistory} />
  ));
}

function AddHistory(props) {
  const { mutate } = props.client;
  const { push, pageNumber } = props;
  mutate({ mutation: gql`
    mutation {
      createPageHistory(number: ${pageNumber}) {
        number
      }
    },
    `, refetchQueries: [{ query: gql`
          {
            allPageHistories(orderBy: id_DESC, first: 5) {
              id
              number
            }
          }
        ` }] }).catch(() => {});
  push(`/history/${pageNumber}`);
  return null;
};

export class History extends React.Component {
  render() {
    return (
      <Query
        query={gql`
          {
            allPageHistories(orderBy: id_DESC, first: 5) {
              id
              number
            }
          }
        `}
      >
        {({ loading, error, data }) => renderQuery(loading, error, data)}
      </Query>
    );
  }
}

export default withApollo(AddHistory as any);

