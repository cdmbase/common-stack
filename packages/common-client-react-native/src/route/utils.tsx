import * as React from 'react';
import { IRoute as IRouteProps, IRouteComponentProps } from '@common-stack/client-react';
import { IScreen,  } from '../interfaces';
import { nanoid } from 'nanoid/non-secure';
import { Redirect } from 'react-router-config';

export function flattenRoutes(routes?: IRouteProps[], parent?: IScreen): IScreen[] {
    if (!Array.isArray(routes)) return [];
    const screens: IScreen[] = [];
    for (let idx = 0; idx < routes.length; idx++) {
        const route = routes[idx];
        const { key: routeKey, path, exact, component, strict, redirect, routes: children, ...options } = route;
        const screenKey = routeKey || nanoid();
        const screen: IScreen = {
            key: screenKey,
            name: path || '/',
            component: function ScreenComponent(props) {
                if (redirect) {
                    return <Redirect from={path} to={redirect} exact={exact} strict={strict} />;
                }
                const children = component ? React.createElement(component, props) : null;
                return parent && parent.component ? React.createElement(parent.component, props, children) : children;
            },
            options: {
                ...options,
                routeMatchOpts: route,
            },
        };
        if (Array.isArray(children) && children.length > 0) {
            screens.push(...flattenRoutes(children, screen));
        } else {
            screens.push(screen);
        }
    }
    return screens;
}

