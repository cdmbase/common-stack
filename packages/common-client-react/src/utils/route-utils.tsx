import * as React from 'react';
import * as Loadable from 'react-loadable';
import { IRouteData } from '../interfaces';
import { matchRoutes } from 'react-router-config';
import * as _ from 'lodash';
import { ConsoleLogger, IConsoleLoggerSettings } from '@cdm-logger/server';
import { LoggerLevel } from '@cdm-logger/core';
import * as Logger from 'bunyan';
export const logger: Logger = ConsoleLogger.create('test', { level: 'trace' });
export const dynamicWrapper = (component: () => any, loading?: any) => Loadable({
    loader: component,
    loading: loading || <div> Loading...</div>,
});




export function getRelation(str1: string, str2: string) {
    if (str1 === str2) {
        console.warn('Two paths are equal');
    }
    const arr1 = str1.split('/');
    const arr2 = str2.split('/');
    if (arr2.every((item, index) => item === arr1[index])) {
        return 1;
    } else if (arr1.every((item, index) => item === arr2[index])) {
        return 2;
    }
    return 3;
}

export function getRenderArr(routes: string[]) {
    let renderArr = [];
    renderArr.push(routes[0]);
    for (let i = 1; i < routes.length; i += 1) {
        let isAdd = false;
        isAdd = renderArr.every(item => getRelation(item, routes[i]) === 3);
        renderArr = renderArr.filter(item => getRelation(item, routes[i]) !== 1);
        if (isAdd) {
            renderArr.push(routes[i]);
        }
    }
    return renderArr;
}

/**
 * Provides the routes based on the index search path.
 * For example, for routerData = {
 * ['/a']: {},
 * ['/a/1]: {},
 * ['/a/2]: {},
 * ['/a/2/1]: {},
 * ['/b/1]: {},
 * ['/b/2]: {},
 * ['/ab/1]: {},
 * }
 *
 * result of getRoutes('/a', routerData) will be
 * [
 * {path: '/a/1', ...},
 * {path: '/a/2', ...},
 * ]
 *
 *
 * @param path
 * @param routerData
 */
export function getRoutes(path: string, routerData: IRouteData) {
    if (path[path.length - 1] !== '/') {
        path += '/'; //  Add a '/' to exclude incomplete paths
    }
    let routes = Object.keys(routerData).filter(routePath => {
        return routePath.indexOf(path) === 0 && routePath !== path;
    });
    // Replace path to '' eg. path='user' /user/name => name.
    routes = routes.map(item => item.replace(path, ''));
    // Conversion and stitching parameters.
    let mergedData = {};
    const renderRoutes = routes.map(item => {
        const exact = !routes.some(route => route !== item && getRelation(route, item) === 1);
        const routeObject = { ...routerData[`${path}${item}`] };

        if (exact) {
            // split and merge
            // if we receive /a/1/2 it should merge as 
            // { path: '/a', routes: [{ path: '/1', routes: [{ path: '/a/1/2'}]}]}
            const arr = (`${path}${item}`).split('/').filter(x => x);
            console.log(`${path}${item}`)
            arr.map((value, i) => {
                console.log(value);
                let newVal = '';
                let newK = {};

                let newMerge: { path?: string, routes?: any[] } = {};
                for (let st = 0; st <= i; st++) {
                    newVal = newVal.concat(`/${arr[st]}`);
                    if (_.isEmpty(newMerge)) {
                        // creates initial object
                        newMerge = _.merge(newMerge, { path: newVal });
                    } else {
                        // probably > first loop
                        // if (sub) {
                        // newK = _.merge(newK, { addToSub(sub, { path: newVal }) });
                        // newK = _.merge(newK, { addToSub(sub, { path: newVal }) });
                        newK = _.merge(newK, { ...addToSub(newMerge, { path: newVal }) });
                        // } else {
                        //     newK = { path: newVal };
                        // }
                        newMerge = _.merge(newMerge, { ...newK });
                    }

                }
                logger.debug('with [%s] == [%j]', newVal, newMerge);
                mergedData = _.merge(mergedData, {});
            });
        }
        return {
            ...routeObject,
            key: `${path}${item}`,
            path: `${path}${item}`,
            exact,
        };
    });
    return renderRoutes;
}

const addToSub = (sub, val) => {
    logger.debug('addToSub %j => %j', val, sub);
    if (sub && sub.routes) {
        logger.debug('...looping..');
        let c = addToSub({ ...sub.routes[0] }, val);
        logger.debug('merge source %j with %j', sub, c);
        sub = _.merge(sub, { routes: [c] });
    } else {
        logger.debug('expecting no router => %j', sub);
        sub = { ...(sub), routes: [val] };
        logger.debug('moidifed by adding %j => %j', val, sub);
    }

    return sub;
};


