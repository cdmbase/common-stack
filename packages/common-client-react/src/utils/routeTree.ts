import { IRouteData } from '../interfaces';

export default class RouteTree implements IRouteData {
	public root: IRouteData;
	public addStatus: boolean;

	constructor(node: IRouteData) {
		this.root = node;
	}

	public traverse(callback: () => any) {
		function walk(node: IRouteData) {
			callback(node);
			node.routes.forEach(walk);
		}

		walk(this.root);
	}

	public addNode(path: String, addlData: Object, parentPath: Object) {
		let newNode = {
			path,
			...addlData,
			routes: [],
		};

		this.addStatus = false;

		if (this.root === null) {
			this.root = newNode;
			return;
		}

		this.traverse(node => {
			if (node.path === parentPath) {
				node.routes.push(newNode);
				this.addStatus = true;
			}
		});
	}
}
