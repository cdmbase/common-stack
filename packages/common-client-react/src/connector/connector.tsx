import * as React from 'react';
import {Route, Switch} from 'react-router-dom';
import {AbstractFeature} from '@common-stack/client-core';
import * as Logger from 'bunyan';
import {IPlugin, IReactFeature} from '../interfaces';
import {getMenus, getRoutes} from '../utils';
import {castArray, map, union, without} from 'lodash';
import {registerPlugin, getPlugins, getPlugin} from '../plugin-area/plugin-api';

const combine = (features, extractor) => without(union(...map(features,
        res => castArray(extractor(res)))), undefined);

export class Feature extends AbstractFeature implements IReactFeature {
    private logger;
    public componentFillPlugins;
    constructor(
        feature?: IReactFeature,
        // tslint:disable:trailing-comma
        ...features: IReactFeature[]
    ) {
        super(feature, ...features);
        this.logger = Logger.createLogger({
            name: Feature.name
        });
        this.componentFillPlugins = this.registerComponentFillPlugins(combine(arguments,
            (arg: IReactFeature) => arg.componentFillPlugins));
    }

    /* tslint:disable:jsx-no-lambda */
    private renderRoutes = (routes, solidRoutes, extraProps = {}, switchProps = {}) =>
        routes ? (
            <Switch {...switchProps}>
                {[
                    ...solidRoutes, ...routes.map((route, i) => (
                        <Route
                            key={route.key || i}
                            path={route.path}
                            exact={route.exact}
                            strict={route.strict}
                            render={props => (
                                <route.component {...props} {...extraProps} route={route}/>
                            )}
                        />
                    )),
                ]}
            </Switch>
        ) : null

    /**
     * Get the routes
     */
    public getRoutes(withRoot?: boolean, rootComponent?: any) {
        const configuredRoutes = this.getConfiguredRoutes();
        const solidRoutes = this.route.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {key: component.props.path}));
        return this.renderRoutes(configuredRoutes, solidRoutes);
    }

    /**
     * Get menus
     */
    public getMenus(withRoot?: boolean, rootComponent?: any) {
        return this.getConfiguredMenus();
    }

    /**
     * get configured routes.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredRoutes(searchPath = '/') {
        const routes = Object.assign({}, ...this.routeConfig);
        return getRoutes(searchPath, {...routes});
    }


    /**
     * get configured menus.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredMenus(searchPath = '/') {
        const routes = Object.assign({}, ...this.menuConfig);
        return getMenus(searchPath, {...routes});
    }

    get navItems() {
        return this.navItem.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemsRight.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItems.length,
            }),
        );
    }

    public getWrappedRoot(root: React.ReactNode, req?: any): React.ReactNode {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

    public getDataRoot(root) {
        let nestedRoot = root;
        for (const component of this.dataRootComponent) {
            nestedRoot = React.createElement(component, {}, nestedRoot);
        }
        return nestedRoot;
    }

    public registerLanguages(monaco) {
        for (const func of this.languagesFuncs) {
            func(monaco);
        }
    }

    private registerComponentFillPlugins(plugins): void {
        plugins.forEach((i) => {
            const {name} = i;
            const isFound = !!getPlugin(name);
            if (!isFound) {
                registerPlugin(name, i);
            }
        });
        return plugins;
    }

    public getComponentFillPlugins() {
        return getPlugins();
    }
}
