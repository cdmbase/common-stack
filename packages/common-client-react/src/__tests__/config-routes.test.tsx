import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as Loadable from 'react-loadable';
import { Route } from 'react-router-dom';
import { getRoutes } from '../utils';
import { Feature } from '../connector';
import 'jest';


function waitFor(delay) {
  return new Promise(resolve => {
    setTimeout(resolve, delay);
  });
}

function createLoader(delay, loader, error?: any) {
  return () => {
    return waitFor(delay).then(() => {
      if (loader) {
        return loader();
      } else {
        throw error;
      }
    });
  };
}

function MyLoadingComponent(props) {
  return <div>MyLoadingComponent {JSON.stringify(props)}</div>;
}

function MyComponent(props) {
  return <div>MyComponent {JSON.stringify(props)}</div>;
}


afterEach(async () => {
  try {
    await Loadable.preloadAll();
  } catch (err) {
  }
});

// Test case borrowed from
// https://github.com/jamiebuilds/react-loadable/blob/master/__tests__/test.js
test('loading success', async () => {
  let LoadableMyComponent = Loadable<{ prop: any }, {}>({
    loader: createLoader(400, () => MyComponent),
    loading: MyLoadingComponent,
  });

  let component1 = renderer.create(<LoadableMyComponent prop="foo"/>);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

  let component2 = renderer.create(<LoadableMyComponent prop="bar"/>);

  expect(component2.toJSON()).toMatchSnapshot(); // reload
});

const routerConfig = (namespace = '') => ({
  [namespace + '/a']: {
    component: () => MyComponent,
  },
  ['/a/1']: {
    component: () => MyComponent,
  },
  [namespace + '/a/1']: {
    component: () => MyComponent,
  },
  [namespace + '/a/2']: {
    component: () => MyComponent,
  },
  [namespace + '/a/2/1']: {
    component: () => MyComponent,
  },
  [namespace + '/ab/2/1']: {
    component: () => MyComponent,
  },
  [namespace + '/b/1']: {
    component: () => MyComponent,
  },
  [namespace + '/b/login/register']: {
    component: () => MyComponent,
  },
});

describe('routeConfig getRoutes', () => {


  test('getRoutes based on index path', async () => {
    const result = {
      path: '/a',
      exact: false,
      root: true,
      routes:
        [{ path: '/a/1', exact: true, routes: [] },
          { path: '/a/2', exact: false, routes: [{ 'exact': true, 'path': '/a/2/1', 'routes': [] }] },
          { path: '/ab/2/1', exact: true, routes: [] }],
    };

    const routes = getRoutes('/a', routerConfig());
    expect(routes).toMatchSnapshot();
  });

  test('getRoutes based on index path', async () => {
    const result = {
      path: '/',
      exact: false,
      root: true,
      routes:
        [{
          path: '/a', exact: false, routes: [
            {
              exact: true,
              path: '/a/1',
              routes: [],
            },
            {
              exact: false,
              path: '/a/2',
              routes: [
                {
                  exact: true,
                  path: '/a/2/1',
                  routes: [],
                },
              ],
            },
          ],
        },
          { path: '/ab/2/1', exact: true, routes: [] },
          { path: '/b/1', exact: true, routes: [] },
          { path: '/b/login/register', exact: true, routes: [] }],
    };
    const routes = getRoutes('/', routerConfig());
    expect(routes).toMatchSnapshot();
    expect(JSON.parse(JSON.stringify(routes))).toMatchObject(result);
  });

  test('getRoutes with `@`in namespace', async () => {
    try {
      getRoutes('@namespace', routerConfig('@namespace'));
    } catch (e) {
      expect(e.message).toEqual('Invalid path!');
    }
  });
});


describe('connector configuredRoutes', () => {

  test('getRoutes based on index path', async () => {

    const connector = new Feature({ routeConfig: routerConfig() });
    const result = {
      exact: false, path: '/', root: true,
      routes: [{
        exact: false, path: '/a',
        routes: [{ exact: true, path: '/a/1', routes: [] },
          {
            exact: false, path: '/a/2',
            routes: [
              { exact: true, path: '/a/2/1', routes: [] },
            ],
          }],
      },
        { exact: true, path: '/ab/2/1', routes: [] },
        { exact: true, path: '/b/1', routes: [] },
        { exact: true, path: '/b/login/register', routes: [] }],
    };


    const routes = connector.getConfiguredRoutes();
    expect(routes).toMatchSnapshot();

    expect(JSON.parse(JSON.stringify(routes))).toMatchObject(result);
  });

  test('getRoutes based without any namespace', async () => {

    const connector = new Feature({ routeConfig: routerConfig('@namespace') });
    const result = {
      exact: false, path: '/', root: true,
      routes: [{ exact: true, path: '/a/1', routes: [] }],
    };

    expect(JSON.parse(JSON.stringify(connector.getConfiguredRoutes()))).toMatchObject(result);
  });
});


describe('connector routes', () => {
  const staticRoutes = {
    route: [
      <Route key={'static1'} exact={true} path="/static1" component={MyComponent}/>,
      <Route key={'static2'} exact={true} path="/static2" component={MyComponent}/>,
    ],
  };

  test('check static routes', async () => {
    const connector = new Feature(staticRoutes);

    expect(connector.routes).toMatchSnapshot();
  });

  test('merge static routes and configurable routes', async () => {
    const connector = new Feature({ routeConfig: routerConfig() }, new Feature(staticRoutes));

    expect(connector.routes).toMatchSnapshot();
  });


  test('check configurable routes', async () => {
    const connector = new Feature({ routeConfig: routerConfig() });

    expect(connector.routes).toMatchSnapshot();
  });

});
