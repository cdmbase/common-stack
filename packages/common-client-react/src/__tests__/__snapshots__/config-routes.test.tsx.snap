// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading success 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`loading success 4`] = `
<div>
  MyComponent 
  {"prop":"bar"}
</div>
`;
